homeassistant/components/vicare/fan.py

```diff
@@ -0,0 +1,172 @@
+from PyViCare.PyViCareUtils import (
+    PyViCareInvalidDataError,
+    PyViCareNotSupportedFeatureError,
+    PyViCareRateLimitError,
+)
+from PyViCare.PyViCareVentilationDevice import (
+    VentilationDevice as PyViCareVentilationDevice,
+)
+from requests.exceptions import ConnectionError as RequestConnectionError
+
+from homeassistant.components.fan import FanEntity, FanEntityFeature
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.core import HomeAssistant
+from homeassistant.exceptions import ServiceValidationError
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+from homeassistant.util.percentage import (
+    ordered_list_item_to_percentage,
+    percentage_to_ordered_list_item,
+)
+
+from .const import DEVICE_LIST, DOMAIN
+from .entity import ViCareEntity
+from .types import VentilationMode, VentilationProgram, ViCareDevice
+
+_LOGGER = logging.getLogger(__name__)
+
+ORDERED_NAMED_FAN_SPEEDS = [
+    VentilationProgram.LEVEL_ONE,
+    VentilationProgram.LEVEL_TWO,
+    VentilationProgram.LEVEL_THREE,
+    VentilationProgram.LEVEL_FOUR,
+]
+
+PRESET_MODES = [
+    VentilationMode.PERMANENT,
+    VentilationMode.VENTILATION,
+    VentilationMode.SENSOR_DRIVEN,
+    VentilationMode.SENSOR_OVERRIDE,
+]
+
+
+def _build_entities(
+    device_list: list[ViCareDevice],
+) -> list[ViCareFan]:
+    """Create ViCare fan entities for a device."""
+
+    return [
+        ViCareFan(device.config, device.api, "ventilation")
+        for device in device_list
+        if isinstance(device.api, PyViCareVentilationDevice)
+    ]
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    config_entry: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up the ViCare fan platform."""
+
+    device_list = hass.data[DOMAIN][config_entry.entry_id][DEVICE_LIST]
+
+    async_add_entities(
+        await hass.async_add_executor_job(
+            _build_entities,
+            device_list,
+        )
+    )
+
+
+class ViCareFan(ViCareEntity, FanEntity):
+    """Representation of the ViCare ventilation device."""
+
+    _attr_speed_count = len(ORDERED_NAMED_FAN_SPEEDS)
+    _attr_supported_features = FanEntityFeature.SET_SPEED | FanEntityFeature.PRESET_MODE
+    _attributes: dict[str, Any] = {
+        "active_vicare_mode": None,
+        "active_vicare_program": None,
+        "vicare_modes": None,
+        "vicare_programs": None,
+    }
+
+    def __init__(
+        self,
+        device_config: PyViCareDeviceConfig,
+        device: PyViCareDevice,
+        translation_key: str,
+    ) -> None:
+        """Initialize the fan entity."""
+        super().__init__(device_config, device, translation_key)
+        self._attr_translation_key = translation_key
+
+        with suppress(PyViCareNotSupportedFeatureError):
+            self._attributes["vicare_modes"] = self._api.getAvailableModes()
+        self._attributes["vicare_programs"] = self._api.getAvailablePrograms()
```

Comment:
These are the causes of the effects bdraco is talking about

Reply:
Should this be handed into the constructor instead or rather handled in the `update` function?

Reply:
I maybe can add this as a real entity later on (this applied also to other platforms).